Move getter to the SKPaymentTransaction extension

Juanpe opened the pull request:
Resolves #693

Tiny PR to move the `productIdentifier` getter to the `SKPaymentTransaction` extension.

I've created a `StoreKitExtensions` folder to group all the extensions related to `StoreKit` but I don't know if this looks good to you 🙂

taquitos commented on PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift:
```diff
@@ -0,0 +1,30 @@
+        }
+
+        return anyPayment?.productIdentifier
+    }
```
(New style update) Add a new line before final closing param

taquitos commented on PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift:
```diff
@@ -0,0 +1,30 @@
+import StoreKit
+
+public extension SKPaymentTransaction {
+    /// Although neither `payment` nor `productIdentier` is optional, we must continue
```
(New style update) Add a new line after definition here

taquitos commented on PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift:
```diff
@@ -0,0 +1,30 @@
+    /// Although neither `payment` nor `productIdentier` is optional, we must continue
+    /// checking the nullability to likely fix issue https://github.com/RevenueCat/purchases-ios/issues/279
+    @objc var productIdentifier: String? {
+        let anyPayment: SKPayment? = payment
```
(New style update) rename `anyPayment` to `maybePayment`

Juanpe replied:
I really like that naming convention 👌🏼

taquitos reviewed (commented):
Minor new style things, but 👍

aboedo reviewed (commented):
this is awesome!! 

The only thing I'd add is a specific test for it, which we didn't set up before because the method wasn't accessible from outside. 

It should be easy to cover the happy case, and the other cases could follow an approach like [the one here](https://github.com/RevenueCat/purchases-ios/blob/42ac40844311e0c4100da5370382dc6c1cb200ca/PurchasesTests/Purchasing/PurchasesTests.swift#L2282) to ensure all cases are covered

Juanpe commented on PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift:
```diff
@@ -0,0 +1,30 @@
+public extension SKPaymentTransaction {
+    /// Although neither `payment` nor `productIdentier` is optional, we must continue
+    /// checking the nullability to likely fix issue https://github.com/RevenueCat/purchases-ios/issues/279
+    @objc var productIdentifier: String? {
```
good point 👍🏼

Juanpe commented:
Thanks for the feedback 👌🏼 

I'm going to add some tests to cover all the possible scenarios 👍🏼

Juanpe commented:
I've added some tests to try to cover all scenarios. Please, let me know if you miss any 🙂

BTW, I've updated the code of the getter. Why? When I added the tests, I realized that the payment is `nil` when creating an empty `SKPaymentTransaction`. However, the `productIdentifier` is an empty string when creating an empty string `SKPayment`. So, now, the code has early returns that validate if the payment or product identifier is not defined.

aboedo commented on PurchasesTests/Purchasing/PurchasesTests.swift:
```diff
@@ -2314,6 +2314,41 @@ class PurchasesTests: XCTestCase {
+        self.storeKitWrapper.delegate?.storeKitWrapper(self.storeKitWrapper, updatedTransaction: transaction)
+    }
+    
+    func testNilProductIdentifierIfPaymentIsMissing() {
+        let transaction = SKPaymentTransaction()
+        expect(transaction.rc_productIdentifier).to(beNil())
+    }
+    
+    func testNilProductIdentifierIfPaymentDoesNotHaveProductIdenfier() {
+        let transaction = MockTransaction()
+        transaction.mockPayment = SKPayment()
+        
+        expect(transaction.rc_productIdentifier).to(beNil())
+    }
+    
+    func testProductIdentifierFromAnyTransaction() {
+        let expectedProductIdentifier = "com.product.id1"
+        let product = MockSKProduct(mockProductIdentifier: expectedProductIdentifier)
+        let payment = SKPayment(product: product)
+        let transaction = MockTransaction()
+        transaction.mockPayment = payment
+        
+        expect(transaction.rc_productIdentifier).to(equal(expectedProductIdentifier))
+    }
```
could you move these 3 tests to a new file, `SKPaymentTransaction+ExtensionsTests`? Since they're not related to the `Purchases` class, and the test file is already pretty big

Juanpe replied:
yep, makes more sense

Juanpe replied:
Done ✅

aboedo commented on PurchasesCoreSwift/StoreKitExtensions/SKPaymentTransaction+Extensions.swift:
```diff
@@ -0,0 +1,38 @@
+    /// Although neither `payment` nor `productIdentier` is optional, we must continue
+    /// checking the nullability to likely fix issue https://github.com/RevenueCat/purchases-ios/issues/279
+    @objc var rc_productIdentifier: String? {
+        let maybePayment: SKPayment? = payment
+
+        guard maybePayment != nil else {
```
we're not doing the usual `guard let payment = payment` here to keep an explicit comparison `!= nil`, right?

Juanpe replied:
You're right. TBH, this code is tricky 😅 since we are storing an optional reference to "cheat" to the compiler. The problem is that we want the `productIdentifier`, which is not optional, so if we unwrap the payment reference, we must create an optional reference of the product identifier. However, if we keep the payment reference as optional, we can use it to validate whether the product identifier is nil too.

I don't know if I'm explaining myself 😓...anyway...we could do it, but I'm not sure if we have more benefits than having the current one.

The code will look like this:
```swift
@objc var rc_productIdentifier: String? {
    let maybePayment: SKPayment? = payment

    guard let anyPayment = maybePayment else {
        Logger.appleWarning(Strings.purchase.skpayment_missing_from_skpaymenttransaction)
        return nil
    }
    
    let maybeProductIdentifier: String? = anyPayment.productIdentifier

    guard let productIdentifier = maybeProductIdentifier,
          !productIdentifier.isEmpty else {
        Logger.appleWarning(Strings.purchase.skpayment_missing_product_identifier)
        return nil
    }

    return productIdentifier
}
```

WDYT?

aboedo replied:
🤔 I think the current approach is fine, and we do have comments that explain it, so it should be fine. If any devs try to refactor into a `guard let`, they'll run into a compiler issue right away, and the motivation is explained in the comments, so I don't think it's worth going crazy over it. Just wanted to double-check that this was the case.

Juanpe replied:
Exactly, this is the point 🙂 I don't know if the opinion of another developer who doesn't have the whole context could be useful to be sure that the code is understandable or if we had better modify it 🤔

taquitos replied:
I do not have the whole context and I actually find it slightly confusing. Specifically because we don't unwrap `payment` but we go and use `maybePayment` later in the file. I think we should remove `let maybePayment: SKPayment? = payment` and use `guard let payment = payment else { ` instead of `guard let anyPayment = maybePayment else {`

Juanpe replied:
I understand your point, maybe we could update the comment to explain better what this getter does. The problem is that this code `guard let payment = payment` doesn't compile because `payment` is not optional. So? Why make it more complex than it seems? 😅 

To give you more context on this issue, we are extending the `SKPaymentTransaction` class by adding a shortcut to get the product identifier. This should be something like this: 
```swift
extension SKPaymentTransaction {
    var productIdentifier: String {
        payment.productIdentifier
    }
}
```

But considering issue #279, sometimes `payment` and `productIdentifier` can be `nil`, in this case, we have to treat them as nullable. So the approach I took was to create an optional reference, then the compiler would allow us to check for nullability.

Do you know another way to solve this problem easier? 🤔

taquitos replied:
AHhh, right, ok. I'm a big fan of comments, and I think that would really help. @aboedo explained some the context here around incorrect nullability, so this makes more sense now. So, I'm ok with this + a comment 😄

beylmk replied:
Hey @Juanpe, caught up on this thread and 🤯 . Thanks for taking on a tricky one...

I tried something _slightly_ different out that seems to work:

```suggestion
        guard let maybePayment = payment as SKPayment? else {
            Logger.appleWarning(Strings.purchase.skpayment_missing_from_skpaymenttransaction)
            return nil
        }
        guard let productIdentifier = maybePayment.productIdentifier as String?,
              !productIdentifier.isEmpty else {
            Logger.appleWarning(Strings.purchase.skpayment_missing_product_identifier)
            return nil
        }
```

It's basically the same as your implementation, with the first `maybePayment` bit squashed down. I think it just makes it a little more explicit that `payment` really is of type `SKPayment?` by explicitly casting it, and then doing the same for `productIdentifier`. What do you think?

Juanpe replied:
Hi @beylmk! Thanks for the suggestion, I think it's a tiny change but it reduces the logic, so I will definitely include it. I just have a question related to the swift style guide. It's totally minor, but does it make sense to use `maybePayment` yet? I mean, the prefix `maybe` is used to identify optional `vars`, but if the unwrapping operation is used as the guard condition, the var is not optional. Do you know what I mean? 
 
Some suggestions could be...
```swift
guard let payment = payment as? SKPayment else { ... }

// or

guard let anyPayment = payment as? SKPayment else { ... }

// or

guard let skPayment = payment as? SKPayment else { ... }
```

WDYT? the var should be still called `maybePayment` or we should use another one?  🤔

aboedo replied:
🤔 intersting. My two cents is that: 
```swift
guard let maybePayment = payment as SKPayment? else {
```
indicates that `payment` can safely be interpreted as `SKPayment?` (which makes sense), and we're checking for `nil` with `guard let`, while
```swift
guard let payment = payment as? SKPayment else { ... }
```
kinda reads like we're expecting that `payment` could be of an entirely different type (i.e.: string), and that's what we're checking for. 

I don't have a preference for `maybePayment` vs `payment` in this particular context, but if the cast works correctly with ` as SKPayment?`, I'd go with that. 

That said, I think as long as the _why_ is explained nicely with a comment, either would work. 

@Juanpe I'm happy to go with what you like best here. Thoughts?

Juanpe replied:
yes, you convinced me, the second option might not be completely clear. I prefer to use `maybe` so as to indicating that the `payment` can be `nil`.

The current code uses `maybe` and I added a comment, do you think this is enough or we should add a comment related to this thread?

aboedo replied:
I think the comment is clear enough, and the compiler and tests won't allow us to break this by refactoring it incorrectly, so we're good to go!

aboedo reviewed (commented):
looks great so far! left a couple of small comments

aboedo commented:
re: `guard` and multiple conditions, we don't have anything in the style guide yet, but I usually do
```swift
guard let foo = maybeFoo,
    let bar = maybeBar,
    maybeBar.shouldDoTheThing else {
    ....
}
```

Juanpe commented on Purchases/Public/RCPurchases.m:
```diff
@@ -1245,8 +1245,8 @@ - (void)handlePurchasedTransaction:(SKPaymentTransaction *)transaction {
 - (void)fetchProductsAndPostReceiptWithTransaction:(SKPaymentTransaction *)transaction data:(NSData *)data {
-    if ([self productIdentifierFrom:transaction]) {
-        [self productsWithIdentifiers:@[[self productIdentifierFrom:transaction]]
+    if (transaction.rc_productIdentifier) {
+        [self productsWithIdentifiers:@[transaction.rc_productIdentifier]
```
oups, good catch 👌🏼

Juanpe commented:
Thanks for the feedback! 🤙 I already made the suggested changes. Please, let me know if I miss something 🙂

taquitos reviewed (approved):
🎈🐐  🎉

aboedo reviewed (approved):
good to go as soon as we're happy with the casting thing!

aboedo reviewed (approved):
@Juanpe thanks again for taking care of this one, and for your continued patience while figuring out the trickier parts of the migration!! 🏎
